packages/push/src/instruction/common/swap.rs

```diff
@@ -0,0 +1,164 @@
+            Err(error) => Err(Error::recoverable(state, error)),
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use proptest::{collection::vec, prelude::any};
+    use test_strategy::proptest;
+
+    use super::*;
+    use crate::{
+        instruction::Instruction,
+        push_vm::{push_state::PushState, stack::StackError},
+    };
+
+    #[test]
+    fn swap_empty_stack() {
+        let state = PushState::builder()
+            .with_max_stack_size(0)
+            .with_no_program()
+            .build();
+        let result = Swap::<i64>::new().perform(state).unwrap_err();
+        assert!(result.is_recoverable());
+        assert_eq!(
+            result.error(),
+            &PushInstructionError::StackError(StackError::Underflow {
+                num_requested: 2,
+                num_present: 0
+            })
+        );
+    }
+
+    #[test]
+    fn swap_singleton_stack() {
+        // Check that calling `Swap` on a stack with one element returns a
+        // `StackError::Underflow` error.
+        let state = PushState::builder()
+            .with_max_stack_size(1)
+            .with_int_values([1])
+            .unwrap()
+            .with_no_program()
+            .build();
+        let result = Swap::<i64>::new().perform(state).unwrap_err();
+        assert_eq!(
+            result.error(),
+            &PushInstructionError::StackError(StackError::Underflow {
+                num_requested: 2,
+                num_present: 1
+            })
+        );
+    }
+
+    #[test]
+    fn swap_non_empty_stack() {
+        let state = PushState::builder()
+            .with_max_stack_size(2)
+            .with_int_values([1, 2])
+            .unwrap()
+            .with_no_program()
+            .build();
+        assert_eq!(state.stack::<i64>().top().unwrap(), &1);
+        let result = Swap::<i64>::new().perform(state).unwrap();
+        assert_eq!(result.stack::<i64>().top2().unwrap(), (&2, &1));
+    }
+
+    #[proptest]
+    fn swap_on_stack_with_multiple_items(
+        // We want at least two items so the swap can actually happen.
+        #[strategy(vec(any::<i64>(), 2..100))] mut values: Vec<i64>,
+    ) {
+        let state = PushState::builder()
+            .with_max_stack_size(100)
+            .with_int_values(values.clone().into_iter())
+            .unwrap()
+            .with_no_program()
+            .build();
+        let swap = Swap::<i64>::default();
+        let result = swap.perform(state).unwrap();
+        assert_eq!(result.stack::<i64>().size(), values.len());
+        // We need to swap the first two elements of the set of values
+        // to mimic the swap, and then reverse the set of values since
+        // they are essentially reversed when they are initially pushed
+        // onto the stack.
+        values.swap(0, 1);
+        values.reverse();
+        // Now the set of values of the stack should be the same as `values`.
+        assert_eq!(result.stack::<i64>(), &values);
```

Comment:
we should probably move to this style of assert on stacks more often, instead of using e.g. `.size()` followed by `.top2()`

Reply:
But probably also not a thing we need to do in this pr

Reply:
Totally agreed. I'll make another issue (#356).